parse_regressor.cc can't be compiled with g++

It's been at least 6 month I haven't look into VW code and it changed a lot since when. I was working with my old fork, but today decided to compare it with master branch. And faced with the fact that I can't build master branch with g++ which is default. With libtool or without it.
I'm getting following errors:
```
libtool: compile:  g++ -DHAVE_CONFIG_H -I. -I/usr/include -I/include -Wall -Wno-unused-local-typedefs -pedantic -I ../rapidjson/include -O3 -fomit-frame-pointer -DNDEBUG -MT parse_regressor.lo -MD -MP -MF .deps/parse_regressor.Tpo -c parse_regressor.cc  -fPIC -DPIC -o .libs/parse_regressor.o
parse_regressor.cc: In function ‘void initialize_regressor(vw&, T&)’:
parse_regressor.cc:55:22: error: parse error in template argument list
     weights.template set_default<float,set_initial<T> >(all.initial_weight);
                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
parse_regressor.cc:55:22: error: expected identifier
parse_regressor.cc:57:22: error: parse error in template argument list
     weights.template set_default<random_positive<T> >();
                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
parse_regressor.cc:57:22: error: expected identifier
parse_regressor.cc:59:22: error: parse error in template argument list
     weights.template set_default<random_weights<T> >();
                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
parse_regressor.cc:59:22: error: expected identifier
```
It seems the code is too complicated for g++. My knowledge of templates was far not enough to bypass this. So I switched to clang++ and successfully build the project and passed `make test`.

My compilers are:

```
gcc version 6.2.0 20161005 (Ubuntu 6.2.0-5ubuntu12)
clang version 3.8.1-12ubuntu1 (tags/RELEASE_381/final)
```
Kubuntu 16.10 , 4.8.0-37-generic GNU/Linux
Could someone reproduce this problem? It's most probably g++ version dependent .

I'm having the same issue on Ubuntu 16.10. Using make with CXX unset, that is using gcc version 6.2.0 20161005 (Ubuntu 6.2.0-5ubuntu12), I have the regression error.  Compiling the clang (clang version 3.8.1-12ubuntu1 (tags/RELEASE_381/final)) and CXXFLAGS=" -std=c++11" gives the link error:

```
/bin/bash ../libtool  --tag=CXX   --mode=link clang -I/usr/include -I/include  -Wall -Wno-unused-local-typedefs    -pedantic  -I ../rapidjson/include   -O3 -fomit-frame-pointer -DNDEBUG -std=c++11   -o vw vw-main.o libvw.la liballreduce.la -L/usr/lib/x86_64-linux-gnu -lboost_program_options -L/lib -lz -lpthread
libtool: link: clang -I/usr/include -I/include -Wall -Wno-unused-local-typedefs -pedantic -I ../rapidjson/include -O3 -fomit-frame-pointer -DNDEBUG -std=c++11 -o .libs/vw vw-main.o  ./.libs/libvw.so ./.libs/liballreduce.so -L/usr/lib/x86_64-linux-gnu -lboost_program_options -L/lib -lz -lpthread
/usr/bin/ld: vw-main.o: undefined reference to symbol '_ZNSo9_M_insertIdEERSoT_@@GLIBCXX_3.4.9'
//usr/lib/x86_64-linux-gnu/libstdc++.so.6: error adding symbols: DSO missing from command line
clang: error: linker command failed with exit code 1 (use -v to see invocation)
Makefile:579: recipe for target 'vw' failed
make[2]: *** [vw] Error 1
```

Leaving the CXXFLAG unset produces the following:

```
libtool: compile:  clang -DHAVE_CONFIG_H -I. -I/usr/include -I/include -Wall -Wno-unused-local-typedefs -pedantic -I ../rapidjson/include -O3 -fomit-frame-pointer -DNDEBUG -MT allreduce_sockets.lo -MD -MP -MF .deps/allreduce_sockets.Tpo -c allreduce_sockets.cc  -fPIC -DPIC -o .libs/allreduce_sockets.o
In file included from allreduce_sockets.cc:28:
In file included from ./allreduce.h:40:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/6.2.0/../../../../include/c++/6.2.0/future:35:
/usr/bin/../lib/gcc/x86_64-linux-gnu/6.2.0/../../../../include/c++/6.2.0/bits/c++0x_warning.h:32:2: error: This file requires
      compiler and library support for the ISO C++ 2011 standard. This support must be enabled with the -std=c++11 or -std=gnu++11
      compiler options.
#error This file requires compiler and library support \
 ^
In file included from allreduce_sockets.cc:28:
./allreduce.h:95:8: error: no type named 'mutex' in namespace 'std'
  std::mutex* m_mutex;
  ~~~~~^
./allreduce.h:96:8: error: no type named 'condition_variable' in namespace 'std'
  std::condition_variable* m_cv;
  ~~~~~^
./allreduce.h:225:53: error: use of undeclared identifier 'nullptr'
      { if (max_fd > 0 && select((int)max_fd, &fds, nullptr, nullptr, nullptr) == -1)
                                                    ^
./allreduce.h:225:62: error: use of undeclared identifier 'nullptr'
      { if (max_fd > 0 && select((int)max_fd, &fds, nullptr, nullptr, nullptr) == -1)
                                                             ^
./allreduce.h:225:71: error: use of undeclared identifier 'nullptr'
      { if (max_fd > 0 && select((int)max_fd, &fds, nullptr, nullptr, nullptr) == -1)
                                                                      ^
allreduce_sockets.cc:49:9: error: use of undeclared identifier 'nullptr'
    if (nullptr == inet_ntop(AF_INET, &(far_end.sin_addr), dotted_quad, INET_ADDRSTRLEN))
        ^
allreduce_sockets.cc:111:17: error: use of undeclared identifier 'nullptr'
  if (master == nullptr)
                ^
allreduce_sockets.cc:188:9: error: use of undeclared identifier 'nullptr'
    if (nullptr == inet_ntop(AF_INET, (char*)&parent_ip, dotted_quad, INET_ADDRSTRLEN))
        ^
9 errors generated.
Makefile:690: recipe for target 'allreduce_sockets.lo' failed
make[2]: *** [allreduce_sockets.lo] Error 1
make[2]: Leaving directory '/home/hb55683/work/DecisionService/GIT/vowpal_wabbit/vowpalwabbit'
Makefile:432: recipe for target 'all' failed
make[1]: *** [all] Error 2
make[1]: Leaving directory '/home/hb55683/work/DecisionService/GIT/vowpal_wabbit/vowpalwabbit'
Makefile:570: recipe for target 'all-recursive' failed
make: *** [all-recursive] Error 1
```

Should I open a new issue for Ubuntu 16.10?

--- a/vowpalwabbit/array_parameters.h
+++ b/vowpalwabbit/array_parameters.h
@@ -137,18 +137,18 @@ private:
 
 	inline weight& strided_index(size_t index){ return operator[](index << _stride_shift);}
 
-	template<class R, void(*T)(iterator&, R&)> void set_default(R& info)
+	template<class R, class T> void set_default(R& info)
 	{
 	  iterator iter = begin();
 	  for (size_t i = 0; iter != end(); ++iter, i += stride())
-	    T(iter, info);
+	    T::func(iter, info);
 	}
 
-	template<void(*T)(iterator&)> void set_default()
+	template<class T> void set_default()
 	{ 
 	  iterator iter = begin();
 	  for (size_t i = 0; iter != end(); ++iter, i += stride())
-	    T(iter);
+	    T::func(iter);
 	}
 
 	void set_zero(size_t offset)
@@ -310,15 +310,15 @@ public:
 		_seeded = true;
 	}
 
-	template<class R, void(*T)(iterator&, R&)> void set_default(R& info)
+	template<class R, class T> void set_default(R& info)
 	{
 	  R& new_R = calloc_or_throw<R>();
 	  new_R = info;
 	  default_data = &new_R;
-	  fun = (void(*)(iterator&, void*))T;
+	  fun = (void(*)(iterator&, void*))T::func;
 	}
 
-	template<void(*T)(iterator&)> void set_default() { fun = (void(*)(iterator&, void*))T; }
+	template<class T> void set_default() { fun = (void(*)(iterator&, void*))T::func; }
 
 	void set_zero(size_t offset)
 	{

--- a/vowpalwabbit/gd.cc
+++ b/vowpalwabbit/gd.cc
@@ -824,11 +824,15 @@ void save_load_online_state(vw& all, io_buf& model_file, bool read, bool text, g
 		save_load_online_state(all, model_file, read, text, g, msg, all.weights.dense_weights);
 }
 
-  template<class T> void set_initial_gd(typename T::iterator& iter, pair<float,float>& initial)
+  template<class T> class set_initial_gd_wrapper 
   {
-    (&(*iter))[0] = initial.first;
-    (&(*iter))[1] = initial.second;
-  }
+      public: 
+          static void func(typename T::iterator& iter, pair<float,float>& initial)
+              {
+                (&(*iter))[0] = initial.first;
+                (&(*iter))[1] = initial.second;
+              }
+  };
 
 void save_load(gd& g, io_buf& model_file, bool read, bool text)
 { vw& all = *g.all;
@@ -840,9 +844,9 @@ void save_load(gd& g, io_buf& model_file, bool read, bool text)
       float init_weight = all.initial_weight;
       pair<float,float> p = make_pair(init_weight, all.initial_t);
       if (all.weights.sparse)
-	all.weights.sparse_weights.set_default<pair<float,float>, set_initial_gd<sparse_parameters> >(p);
+	all.weights.sparse_weights.set_default<pair<float,float>, set_initial_gd_wrapper<sparse_parameters> >(p);
       else
-	all.weights.dense_weights.set_default<pair<float,float>, set_initial_gd<dense_parameters> >(p);
+	all.weights.dense_weights.set_default<pair<float,float>, set_initial_gd_wrapper<dense_parameters> >(p);
       //for adaptive update, we interpret initial_t as previously seeing initial_t fake datapoints, all with squared gradient=1
       //NOTE: this is not invariant to the scaling of the data (i.e. when combined with normalized). Since scaling the data scales the gradient, this should ideally be
       //feature_range*initial_t, or something like that. We could potentially fix this by just adding this base quantity times the current range to the sum of gradients

--- a/vowpalwabbit/gd_mf.cc
+++ b/vowpalwabbit/gd_mf.cc
@@ -220,13 +220,17 @@ void mf_train(gdmf& d, example& ec)
 		mf_train(d, ec, d.all->weights.dense_weights);
 }
 
-template <class T>
-void set_rand(typename T::iterator& iter, uint32_t& stride)
+template <class T> class set_rand_wrapper 
 {
-  uint64_t index = iter.index();
-  for (weight_iterator_iterator w = iter.begin(); w != iter.end(stride); ++w, ++index)
-    *w = (float)(0.1 * merand48(index));
-}
+public:
+    
+    static void func(typename T::iterator& iter, uint32_t& stride)
+    {
+      uint64_t index = iter.index();
+      for (weight_iterator_iterator w = iter.begin(); w != iter.end(stride); ++w, ++index)
+        *w = (float)(0.1 * merand48(index));
+    }
+};
 
 void save_load(gdmf& d, io_buf& model_file, bool read, bool text)
 { vw& all = *d.all;
@@ -237,9 +241,9 @@ void save_load(gdmf& d, io_buf& model_file, bool read, bool text)
 	{
 	  uint32_t stride = all.weights.stride();
 	  if (all.weights.sparse)
-	    all.weights.sparse_weights.set_default<uint32_t, set_rand<sparse_parameters> >(stride);
+	    all.weights.sparse_weights.set_default<uint32_t, set_rand_wrapper<sparse_parameters> >(stride);
 	  else
-	    all.weights.dense_weights.set_default<uint32_t, set_rand<dense_parameters> >(stride);
+	    all.weights.dense_weights.set_default<uint32_t, set_rand_wrapper<dense_parameters> >(stride);
 	}
     }
 

--- a/vowpalwabbit/lda_core.cc
+++ b/vowpalwabbit/lda_core.cc
@@ -668,11 +668,13 @@ struct initial_weights
     : _initial(initial), _initial_random(initial_random), _random(random), _lda(lda), _stride(stride){}
 };
 
-template<class T>
-void set_initial_lda(typename T::iterator& iter, initial_weights& iw)
+template<class T> class set_initial_lda_wrapper
 {
-	uint32_t lda = iw._lda;
-	weight initial_random = iw._initial_random;
+public:    
+    static void func(typename T::iterator& iter, initial_weights& iw)
+    {
+    	uint32_t lda = iw._lda;
+    	weight initial_random = iw._initial_random;
 	if (iw._random)
 	  {
 	    uint64_t index = iter.index();
@@ -682,7 +684,8 @@ void set_initial_lda(typename T::iterator& iter, initial_weights& iw)
 		}
 	  }
 	(&(*iter))[lda] = iw._initial;
-}
+    }
+};
 
 void save_load(lda &l, io_buf &model_file, bool read, bool text)
 { vw& all = *(l.all);
@@ -691,9 +694,9 @@ void save_load(lda &l, io_buf &model_file, bool read, bool text)
   { initialize_regressor(all);
     initial_weights init(all.initial_t, (float)(l.lda_D / all.lda / all.length() * 200), all.random_weights, all.lda, all.weights.stride());
     if (all.weights.sparse)
-      all.weights.sparse_weights.set_default<initial_weights, set_initial_lda<sparse_parameters> >(init);
+      all.weights.sparse_weights.set_default<initial_weights, set_initial_lda_wrapper<sparse_parameters> >(init);
     else
-      all.weights.dense_weights.set_default<initial_weights, set_initial_lda<dense_parameters> >(init);
+      all.weights.dense_weights.set_default<initial_weights, set_initial_lda_wrapper<dense_parameters> >(init);
   }
   if (model_file.files.size() > 0)
   { uint64_t i = 0;

--- a/vowpalwabbit/parse_regressor.cc
+++ b/vowpalwabbit/parse_regressor.cc
@@ -25,19 +25,31 @@ using namespace std;
 #include "vw_validate.h"
 #include "vw_versions.h"
 
-template <class T> void set_initial(typename T::iterator& iter, float& initial) { *iter = initial; }
+template <class T> class set_initial_wrapper 
+{
+public:
+    static void func(typename T::iterator& iter, float& initial) { *iter = initial; }    
+};
 	
-template <class T> void random_positive(typename T::iterator& iter)
+template <class T> class random_positive_wrapper 
 {
-  uint64_t index = iter.index();
-  *iter = (float)(0.1 * merand48(index));
-}
+public:
+   static void func(typename T::iterator& iter)
+    {
+      uint64_t index = iter.index();
+      *iter = (float)(0.1 * merand48(index));
+    }
+};
 
-template <class T> void random_weights(typename T::iterator& iter)
+template <class T> class random_weights_wrapper
 {
-  uint64_t index = iter.index();
-  *iter = (float)(merand48(index) - 0.5);
-}
+public:    
+    static void func(typename T::iterator& iter)
+    {
+      uint64_t index = iter.index();
+      *iter = (float)(merand48(index) - 0.5);
+    }
+};
 
 template<class T> void initialize_regressor(vw& all, T& weights)
 { // Regressor is already initialized.
@@ -52,11 +64,11 @@ template<class T> void initialize_regressor(vw& all, T& weights)
   if (weights.mask() == 0)
     { THROW(" Failed to allocate weight array with " << all.num_bits << " bits: try decreasing -b <bits>"); }
   else if (all.initial_weight != 0.)
-    weights.template set_default<float,set_initial<T> >(all.initial_weight);
+    weights.template set_default<float,set_initial_wrapper<T> >(all.initial_weight);
   else if (all.random_positive_weights)
-    weights.template set_default<random_positive<T> >();
+    weights.template set_default<random_positive_wrapper<T> >();
   else if (all.random_weights)
-    weights.template set_default<random_weights<T> >();
+    weights.template set_default<random_weights_wrapper<T> >();
 }
 
 void initialize_regressor(vw& all)

